### PR TITLE
fix(vanilla): unexpected null state update behavior

### PR DESCRIPTION
**Pull request by wulimao49**
unexpected behavior when update state with [`null`]

[current logic here](https://github.com/pmndrs/zustand/blob/main/src/vanilla.ts#L76C28-L76C28)

the code above will cause something like
```typescript
const useStore = create<string | null>(() => 'default');

useStore.setState(null); // will actually run: Object.assign({}, 'default', null)

useStore.getState(); // get { "0": "d", "1": "e", "2": "f", "3": "a", "4": "u", "5": "l",  "6": "t" }
```

### CODE REVIEW

File: src/vanilla.ts [hunk 1]
```diff
@@ -73,7 +73,7 @@ const createStoreImpl: CreateStoreImpl = (createState) => {
     if (!Object.is(nextState, state)) {
       const previousState = state
       state =
-        replace ?? typeof nextState !== 'object'
+        replace ?? ( typeof nextState !== 'object' || nextState === null )
```
Comment (dai-shi) [inline src/vanilla.ts:76]: Yeah, that makes sense.

**Review by dai-shi — COMMENTED** [review]
Please check CI errors.

File: src/vanilla.ts (hunk 1 above)
Comment (dbritto-dev) [inline src/vanilla.ts:76]: I would use `Object.prototype.toString` since typeof could return `object` in many cases. Ref: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/typeof

```suggestion
        replace ?? Object.prototype.toString.call(nextState).slice(8, -1) === 'Object'
```

**Review by dbritto-dev — CHANGES_REQUESTED** [review]

File: src/vanilla.ts (hunk 1 above)
Comment (dai-shi) [reply src/vanilla.ts:76]: Sorry, but I don't prefer that. typeof should be good for our purpose and for edge cases, users can use `replace`.

Comment (dbritto-dev) [reply src/vanilla.ts]: @dai-shi in that case we need to add a new test case for this change

Comment (dai-shi) [reply src/vanilla.ts]: In any case, a test would be good. @wulimao49 would you be able to add one?

Wait, on second thought, this isn't an issue with `null` specifically.
Maybe, it's more reasonable to do this: `typeof state !== 'object' || state === null`
What do you guys think?

Comment (wulimao49) [reply src/vanilla.ts]: `typeof state !== 'object' || state === null` will make it impossible to set a object state to `null` except explicitly set `replace` to `true`.

both do it on `state` and `nextState` will have some kind of edge cases.

maybe we should make the merging logic configable? or just using `replace` is enough.

 yes i can set a test if we decide a certain solution on this.

Comment (dai-shi) [reply src/vanilla.ts]: Here's my decision. This doesn't cover edge cases. It's intentional.
Just pick either one please:
- `typeof nextState !== 'object' || nextState === null`
- `typeof state !== 'object' || state === null`

Which makes less error-prone?

Comment (dbritto-dev) [reply src/vanilla.ts]: @dai-shi in that case I'll go with the second one

Comment (dai-shi) [reply src/vanilla.ts]: @wulimao49 How about you?

Comment (wulimao49) [reply src/vanilla.ts]: @dai-shi i prefer the second one

Comment (dai-shi) [reply src/vanilla.ts]: Let's go with `typeof state !== 'object' || state === null`.
Can you create some test cases?

Comment (wulimao49) [reply src/vanilla.ts]: test cases added

File: src/vanilla.ts [hunk 2]
```diff
@@ -73,7 +73,7 @@ const createStoreImpl: CreateStoreImpl = (createState) => {
     if (!Object.is(nextState, state)) {
       const previousState = state
       state =
-        replace ?? typeof nextState !== 'object'
+        replace ?? (typeof nextState !== 'object' || nextState === null)
```
Comment (dai-shi) [inline src/vanilla.ts:76]: This is what we agreed, right?
```suggestion
        replace ?? (typeof state !== 'object' || state === null)
```

File: src/vanilla.ts (hunk 1 above)
Comment (dai-shi) [reply src/vanilla.ts]: Yeah, that's too much. 99% users don't use this capability anyway.

I considered changing to `replace ?? defaultReplace`, and `const defalutReplace = typeof state !== 'object || state === null` on initialization. But, its capability is rather limited, and the behavior can be less predictable.

I think the current solution is a fair point.
